Add support for ATECC508A

michpappas opened the pull request:
This series performs the changes required in libs96at to support multiple devices and adds support for the ATECC508A.

michpappas committed:
Add opcodes and status codes for ATECC508A

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Update core command parameters and add ATECC508A execution times

Replace the io_interface parameter in core command functions with the
s96at descriptor. This contains the chip type being currently used,
and allows chip-specific behavior customiazation.

Add command execution times for ATECC508A and update get_command() to
update execution time based on the device used.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Add support for multiple devices in i2c_linux

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Rename core command functions

Rename core command functions to match the command name as specified
in the device's datasheet.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add Info command

Add cmd_info() that implements the Info command.

Info mode 0x00 returns the device revision. In atsha204a, revision
is obtained by the DevRev command. Update s96at_get_devrev() to
use cmd_info() if the device is an atecc508a.

Info mode 0x01 returns whether a slot contains a valid ECC key.
Add s96at_key_valid() that implements the above.

Info mode 0x02 returns information related to the device's dynamic
state. Add s96at_state(), that implements the above.

Info mode 0x03 relates to SWI devices, which does not apply to
Secure96. No function is implemented to retrieve that info.

Update tests.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Update tests to support multiple devices

Add command line option to select device.

Update device numbers in s96at.h to binary flags. This makes it
possible to select among things that apply to either or both
devices.

Add new field to test structure, to specify the device a
given test should be executed on.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Rearrange core commands in alphabetical order

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Support variable slot lengths for atecc508a

Update the s96at_read_data() and s96at_write_data() commands to
specify the address in terms of [slot, block, offset]. This
provides the ability to access slots of different lengths, as
required to support the atecc508a.

Update the API documentation and tests.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Update Configuration zone access functions for atecc508a

In atsha204a, the Configuration zone is written in words, ie
each write is 4 Bytes long. In atecc508a, the Configuration
zone is written in blocks, ie 32 Bytes at a time. Update
s96_write_config() and its documentation to reflect the above.

Reading from the Configuration zone is performed in multiples of
word size in both devices. Update the documentation accordingly.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add GenKey command

Add cmd_genkey() to support the GenKey command.
Add s96at_gen_key() to generate a new EC Private Key, or an EC
Public Key from a Private Key stored into the device, or a digest
of a Public Key stored into the device.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Remove unused function

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Update the max execution time of Nonce on ATECC508A

Update the maximum execution time of Nonce on atecc508a
to the value of atsha204a, as execution would not complete
on time when using the datasheet value.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Update SHA functions for ATECC508A

ATECC508A SHA command has a different interface compared to ATSHA204A.
Update s96at_sha() to support both interfaces. Update tests.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add PrivWrite command

Add cmd_privwrite() to implement the PrivWrite command.
Add s96at_privwrite() to write externally generated ECC keys into
the device.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add Counter command

Add cmd_counter() to implement the Counter command.
Add s96at_get_counter() and s96at_increment_counter() to retrieve and
increment the values of the monotinic counters.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add Sign command

Add cmd_sign() implement the Sign command.

Add s96at_sign() to generate ECDSA signatures from messages generated
internally or externally to the device.

Increase the minimum execution time of the Sign command, as the command
would not complete execution during the time specified in the datasheet.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add Verify command

Add cmd_verify() to support the Verify command.
Add s96at_verify_key() to validate / invalidate EC public keys.
Add s96at_verify_sig() to verify ECDSA signatures.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
ATECC508A: Add ECDH command

Add cmd_ecdh() to support the ECDH command.
Add s96at_edch() to support shared key computation using ECDH.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Correct the return value of cmd_write()

Correct the return value of cmd_write() to the status
carried into the response of the subsequent Read command,
rather than the return status of the Read itself.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
Disable DEBUG flags in CMakeLists.txt

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas commented:
Hi @jbech-linaro, sorry for the rather large patch set, I've been working on this one on and off for a while. There's still a few small bits missing, yet all major functionality is there. A few more examples added in https://github.com/michpappas/secure96_examples. Now with both devices working, it is possible to demonstrate some real-world use-cases :-)

jbech-linaro commented:
Hi @michpappas , sorry for the long delay. Today I'm back from a long vacation and a business trip. I've had a quick look at the patches and so far I think it is looking good. However, I'd like to spend a couple of more minutes looking into this. Are there any updates you'd like to add to the PR? I'm asking since it's been here for almost 5 weeks without anyone looking at it.

michpappas commented:
Hi @jbech-linaro, no further updates from my part, thanks!

jockebech reviewed (commented):
Minor comment, do a `s/atecc508a/ATECC508A/g` in all comments (same for the 204). Otherwise mostly some minor comments. This is good stuff! Thanks once again!

michpappas commented on s96at/include/s96at.h:
```diff
@@ -49,8 +74,35 @@
+
 enum s96at_device {
-	S96AT_ATSHA204A,
+	S96AT_ATSHA204A = 1,
```
Because it allows me to do bitwise operations, eg
`{"Read: OTP", test_read_otp, S96AT_ATSHA204A | S96AT_ATECC508A}`

michpappas commented on s96at/include/s96at.h:
```diff
@@ -153,6 +243,32 @@ uint16_t s96at_crc(const uint8_t *buf, size_t buf_len, uint16_t current_crc);
 
+/* Performs ECDH
+ *
+ * This function is only available on ATECC508A.
```
The thing is ATECC508A is backwards compatible with ATSHA204A, so common.h would include the entire command set of ATSHA204A.

jbech-linaro replied:
Right, I was more thinking something like
```
atsha204a.h: Only things for 204a
atecc508a.h: Only things for the 508a
common.h: Things that are shared between the two
```
Probably `common.h` would be the bigger one and much less in the other two h-files. But as mentioned, if that is of interest we can do that later on. Nothing we have to do now.

michpappas commented on s96at/include/s96at.h:
```diff
@@ -153,6 +243,32 @@ uint16_t s96at_crc(const uint8_t *buf, size_t buf_len, uint16_t current_crc);
+ * This function is only available on ATECC508A.
+ *
+ * Reads the value of the specified counter into the val parameter.
+ * Valid counter values are 0 and 1.
```
These refer to the two counters supported by the device. I updated the documentation to make that more clear.

jbech-linaro replied:
Thanks!

michpappas commented on s96at/src/cmd.c:
```diff
@@ -25,163 +27,267 @@ uint8_t SLOT_CONFIG_ADDR(uint8_t slotnbr)
+	struct cmd_packet p;
+
+	get_command(desc->dev, &p, OPCODE_ECDH);
+	p.param1 = 0;
```
Fixed

michpappas commented on s96at/src/cmd.c:
```diff
@@ -25,163 +27,267 @@ uint8_t SLOT_CONFIG_ADDR(uint8_t slotnbr)
+	get_command(desc->dev, &p, OPCODE_ECDH);
+	p.param1 = 0;
 	p.param2[0] = slotnbr;
+	p.param2[1] = 0;
```
Fixed

michpappas commented on s96at/tests/CMakeLists.txt:
```diff
@@ -9,6 +9,6 @@ add_executable(${PROJECT_NAME} EXCLUDE_FROM_ALL ${SRC} tests.c)
 target_compile_definitions(${PROJECT_NAME}
 	PRIVATE -DI2C_DEVICE="${I2C_DEVICE}"
-	PRIVATE -DDEBUG
+#	PRIVATE -DDEBUG
```
Okay, makes sense. Will write a patch for that too

michpappas commented on s96at/src/s96at.c:
```diff
@@ -394,106 +497,156 @@ uint8_t s96at_get_serialnbr(struct s96at_desc *desc, uint8_t *buf)
-	if (buf_len <= msg_len)
-		return S96AT_STATUS_BAD_PARAMETERS;
+	/* ATECC508A requires a message of at least 64 bytes, and expects the message
+	 * to be passed raw. ATECC204A requires that the SHA-256 padding is applied to
```
Fixed

michpappas commented:
Hi @jbech-linaro, hope you're well. It took a while to look at this one, as I recently relocated to Germany and had to deal of all sorts of things %-) I'll probably won't be able to dedicate the same amount of time on this one as before, but I'd like to know whether there are any ideas of what could come out of this. Perhaps a demo with hardware crypto support on OPTEE would be nice?

jbech-linaro commented:
> Hi @jbech-linaro, hope you're well. It took a while to look at this one, as I recently relocated to Germany and had to deal of all sorts of things %-) I'll probably won't be able to dedicate the same amount of time on this one as before,

I see, well you don't have to come up with any excuses. I'm more than thankful for all the good work you've been doing with this project. You've taken it from a hack to be something really good!

> but I'd like to know whether there are any ideas of what could come out of this. Perhaps a demo with hardware crypto support on OPTEE would be nice?

Yeah, that was what I was thinking about initially too, i.e. one could create a Trusted Application in OP-TEE that have the same kind of functionality as the IC's one Secure96 (one would also need to create a plain C user space application responsible of doing the communication between the IC's and the TEE). Alternatively one could focus on just doing a subset of the features, for example, leverage the signing of data (using 508A) being sent to the TEE. My challenge here on my side is that I have almost no time to spare on a project like that for the moment (and the future doesn't look too bright). One idea would be to eventually try do something as we are getting closer to the next Linaro Connect (March 19') and see it'd be possible to demo something there. Even though I don't have time right now, I'd like to keep the idea open. 

Btw, based on your recent feedback from yesterday, it looks like we can go ahead and merge this patch set, i.e., no planned fixed for this particular patch set, right?

michpappas committed:
Update device names in comments with uppercase

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas commented:
> I see, well you don't have to come up with any excuses. I'm more than thankful for all the good work you've been doing with this project. You've taken it from a hack to be something really good!

Thank you, I enjoyed working on this one.
 
> Yeah, that was what I was thinking about initially too, i.e. one could create a Trusted Application in OP-TEE that have the same kind of functionality as the IC's one Secure96 (one would also need to create a plain C user space application responsible of doing the communication between the IC's and the TEE). Alternatively one could focus on just doing a subset of the features, for example, leverage the signing of data (using 508A) being sent to the TEE. My challenge here on my side is that I have almost no time to spare on a project like that for the moment (and the future doesn't look too bright). One idea would be to eventually try do something as we are getting closer to the next Linaro Connect (March 19') and see it'd be possible to demo something there. Even though I don't have time right now, I'd like to keep the idea open.

Sounds good. I'll have a closer look and get back to you on this one. I think March sounds feasible, wrt my current commitments.

> Btw, based on your recent feedback from yesterday, it looks like we can go ahead and merge this patch set, i.e., no planned fixed for this particular patch set, right?

Yes please go ahead. I updated the comment about the monotonic counters, which I though I did on the previous update and also added one more commit that capitalizes the device names in comments as per your suggestion, but nothing else to do on this patch-set.

jbech-linaro commented:
Merged, thanks @michpappas !
